### PR TITLE
[Feature] YOLOv8 supports using mask annotation to optimize bbox

### PR DESCRIPTION
**Pull request by Nioolek**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

YOLOv8 supports using mask annotation to optimize bbox. 

- [x] LoadAnnotations
- [x] Mosaic
- [x] YOLOv5RandomAffine
- [x] LetterResize
- [x] YOLOv5CopyPaste

### CODE REVIEW

File: configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco.py [hunk 1]
```diff
@@ -11,10 +11,10 @@
 widen_factor = 0.5
 max_epochs = 500
 save_epoch_intervals = 10
-train_batch_size_per_gpu = 16
-train_num_workers = 8
+train_batch_size_per_gpu = 2
+train_num_workers = 1
 val_batch_size_per_gpu = 1
-val_num_workers = 2
+val_num_workers = 1
```
Comment (PeterH0323) [inline configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco.py:17]: 这几个修改不需要提交

File: configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco_copypaste.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
+_base_ = '../_base_/default_runtime.py'
```
Comment (PeterH0323) [inline configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco_copypaste.py:1]: 这个 config 可以继承吗？

Comment (Nioolek) [reply configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco_copypaste.py:1]: > 

这个是测试代码，等测试完毕会删除

File: configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco.py [hunk 3]
```diff
@@ -113,7 +113,7 @@
 
 pre_transform = [
```
Comment (hhaAndroid) [inline configs/yolov8/yolov8_s_syncbn_fast_8xb16-500e_coco.py:114]: yolov8_s_syncbn_fast_8xb16-500e_coco.py -> yolov8_s-mask_refine_syncbn_fast_8xb16-500e_coco.py

File: mmyolo/datasets/transforms/mix_img_transforms.py [hunk 4]
```diff
@@ -384,12 +408,17 @@ def mix_img_transform(self, results: dict) -> dict:
             mosaic_bboxes = mosaic_bboxes[inside_inds]
             mosaic_bboxes_labels = mosaic_bboxes_labels[inside_inds]
             mosaic_ignore_flags = mosaic_ignore_flags[inside_inds]
+            print('!!!', len(results['gt_masks']), inside_inds)
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/mix_img_transforms.py:411]: delete

File: mmyolo/datasets/transforms/transforms.py [hunk 5]
```diff
@@ -364,6 +442,74 @@ class LoadAnnotations(MMDET_LoadAnnotations):
...
 
+    def __init__(self,
+                 mask2bbox: bool = False,
+                 poly2mask: bool = False,
+                 **kwargs) -> None:
+        super().__init__(poly2mask=poly2mask, **kwargs)
+        self.mask2bbox = mask2bbox
+
+    def transform(self, results: dict) -> dict:
+        results = super().transform(results)
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:453]: 这里可以优化下的。如果开了 mask2bbox，那么就没有必要运行一遍 load_bbox 了

File: mmyolo/datasets/transforms/mix_img_transforms.py [hunk 6]
```diff
@@ -314,9 +314,13 @@ def mix_img_transform(self, results: dict) -> dict:
             results (dict): Updated result dict.
         """
         assert 'mix_results' in results
+        assert len(results['gt_masks']) == len(results['gt_bboxes']) == len(
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/mix_img_transforms.py:317]: 这部分代码应该可以直接全部抄mmdet的，已经实现了

File: configs/yolov8/yolov8_m_mask-refine_copypaste_syncbn_fast_8xb16-500e_coco.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+_base_ = './yolov8_s_mask-refine_syncbn_fast_8xb16-500e_coco.py'
```
Comment (hhaAndroid) [inline configs/yolov8/yolov8_m_mask-refine_copypaste_syncbn_fast_8xb16-500e_coco.py:1]: copypaste 名称不需要写到配置名里面去

File: mmyolo/datasets/transforms/transforms.py [hunk 8]
```diff
@@ -558,32 +630,143 @@ def transform(self, results: dict) -> dict:
...
                 valid_index]
             results['gt_ignore_flags'] = results['gt_ignore_flags'][
                 valid_index]
 
-            if 'gt_masks' in results:
-                raise NotImplementedError('RandomAffine only supports bbox.')
         return results
 
+    @staticmethod
+    def warp_poly(poly: np.ndarray, warp_matrix: np.ndarray, img_w: int,
```
Comment (RangiLyu) [inline mmyolo/datasets/transforms/transforms.py:679]: You can refer to `bbox_project` in `mmdet/structures/bbox/transforms.py` to write a `mask_project` function.

File: mmyolo/datasets/transforms/transforms.py [hunk 9]
```diff
@@ -558,32 +630,143 @@ def transform(self, results: dict) -> dict:
...
+                    np.zeros(6, dtype=poly_per_obj[0].dtype)
+                ]
+            new_masks.append(warpped_poly_per_obj)
+
+        gt_masks = PolygonMasks(new_masks, img_h, img_w)
+        return gt_masks
+
+    @staticmethod
+    def resample_masks(gt_masks: PolygonMasks,
+                       resample_num: int = 1000) -> PolygonMasks:
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:746]: resample_num 参数应该作为输入参数，可以修改

File: configs/yolov8/README.md [hunk 10]
```diff
@@ -20,19 +20,25 @@ YOLOv8-P5 model structure
...
-| YOLOv8-m |  P5  | 640  |  Yes   | Yes |   7.2    |  49.8  | [config](https://github.com/open-mmlab/mmyolo/blob/dev/configs/yolov8/yolov8_m_syncbn_fast_8xb16-500e_coco.py) | [model](https://download.openmmlab.com/mmyolo/v0/yolov8/yolov8_m_syncbn_fast_8xb16-500e_coco/yolov8_m_syncbn_fast_8xb16-500e_coco_20230115_192200-c22e560a.pth) \| [log](https://download.openmmlab.com/mmyolo/v0/yolov8/yolov8_m_syncbn_fast_8xb16-500e_coco/yolov8_m_syncbn_fast_8xb16-500e_coco_20230115_192200.log.json) |
+| Backbone | Arch | size | Mask Refine | Copy Paste | SyncBN | AMP | Mem (GB) | box AP |                                 Config                                  |        Download        |
+| :------: | :--: | :--: | :---------: | :--------: | :----: | :-: | :------: | :----: | :---------------------------------------------------------------------: | :--------------------: |
```
Comment (hhaAndroid) [inline configs/yolov8/README.md:24]: Copy Paste 这一列不需要

File: mmyolo/datasets/transforms/transforms.py [hunk 11]
```diff
@@ -1079,3 +1311,183 @@ def _get_valid_inds(self, gt_bbox: HorizontalBoxes,
...
+        copypaste_bboxes = gt_bboxes_flip[valid_inds]
+        gt_bboxes = gt_bboxes.cat([gt_bboxes, copypaste_bboxes])
+
+        # prepare images
+        copypaste_gt_masks = gt_masks[valid_inds]
+        copypaste_gt_masks_flip = copypaste_gt_masks.flip()
+        # convert poly format to bitmap format
+        # example: poly: [[array(0.0, 0.0, 10.0, 0.0, 10.0, 10.0, 0.0, 10.0]]
+        #  -> bitmap: a mask with shape equal to (1, img_h, img_w)
+        copypaste_gt_masks_bitmap = copypaste_gt_masks.to_ndarray()
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:1406]: Compare and analyze whether this line of code causes the training speed to drop

File: mmyolo/datasets/transforms/transforms.py (hunk 8 above)
Comment (Nioolek) [reply mmyolo/datasets/transforms/transforms.py:684]: For quick development purposes, this function will remain the current implementation for now. I will give PR to mmdet to implement the relevant function

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**